src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs

```diff
@@ -273,37 +271,54 @@ private void HandleRemovedDocuments(RequestContext context, ImmutableArray<Previ
+        private ImmutableArray<LSP.Diagnostic> ConvertDiagnostic(IDiagnosticSource diagnosticSource, DiagnosticData diagnosticData, ClientCapabilities capabilities)
         {
-            Contract.ThrowIfNull(diagnosticData.Message, $"Got a document diagnostic that did not have a {nameof(diagnosticData.Message)}");
-
             var project = diagnosticSource.GetProject();
+            var diagnostic = CreateLspDiagnostic(diagnosticData, project, capabilities);
 
-            if (!capabilities.HasVisualStudioLspCapability())
+            // Check if we need to handle the unnecessary tag (fading).
+            if (!diagnosticData.CustomTags.Contains(WellKnownDiagnosticTags.Unnecessary))
             {
-                var diagnostic = CreateBaseLspDiagnostic();
-                return diagnostic;
+                return ImmutableArray.Create<LSP.Diagnostic>(diagnostic);
             }
-            else
+
+            // DiagnosticId supports fading, check if the corresponding VS option is turned on.
+            if (!SupportsFadingOption(diagnosticData))
             {
-                var vsDiagnostic = CreateBaseLspDiagnostic();
-                vsDiagnostic.DiagnosticType = diagnosticData.Category;
-                vsDiagnostic.Projects = new[]
-                {
-                    new VSDiagnosticProjectInformation
-                    {
-                        ProjectIdentifier = project.Id.Id.ToString(),
-                        ProjectName = project.Name,
-                    },
-                };
+                return ImmutableArray.Create<LSP.Diagnostic>(diagnostic);
+            }
 
-                return vsDiagnostic;
+            // Check to see if there are specific locations marked to fade.
+            if (!diagnosticData.TryGetUnnecessaryDataLocations(out var unnecessaryLocations))
+            {
+                // There are no specific fading locations, just mark the whole diagnostic span as unnecessary.
+                diagnostic.Tags = diagnostic.Tags?.Append(DiagnosticTag.Unnecessary);
+                return ImmutableArray.Create<LSP.Diagnostic>(diagnostic);
             }
 
-            // We can just use VSDiagnostic as it doesn't have any default properties set that
-            // would get automatically serialized.
-            LSP.VSDiagnostic CreateBaseLspDiagnostic()
+            // Roslyn produces unnecessary diagnostics by using additional locations, however LSP doesn't support tagging
+            // additional locations separately.  Instead we just create multiple hidden diagnostics for unnecessary squiggling.
+            using var _ = ArrayBuilder<LSP.Diagnostic>.GetInstance(out var diagnosticsBuilder);
+            diagnosticsBuilder.Add(diagnostic);
+            foreach (var location in unnecessaryLocations)
             {
+                var additionalDiagnostic = CreateLspDiagnostic(diagnosticData, project, capabilities);
```

Comment:
Do we also need to change the Severity to Hidden for this additionalDiagnostic?

Reply:
update - LSP doesn't have a hidden severity, instead the visibility is controlled by the tags.  I changed it to the lowest severity anyway